api/src/main/java/org/apache/iceberg/types/Type.java

```diff
@@ -112,6 +113,23 @@ public PrimitiveType asPrimitiveType() {
     Object writeReplace() throws ObjectStreamException {
       return new PrimitiveHolder(toString());
     }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof PrimitiveType)) {
+        return false;
+      }
+
+      PrimitiveType that = (PrimitiveType) o;
+      return typeId() == that.typeId();
```

Comment:
What about `PrimitiveType` where we have an attribute?
Like:
- `TimestampType(boolean adjustToUTC)`
- `FixedType ofLength(int length)`
- `DecimalType of(int precision, int scale)`

Do we consider them equals if the parameters are different?

Reply:
@pvary `TimestampType`, `FixedType` and `DecimalType` have overridden `equals` and `hashCode` methods of `Object` and `PrimitiveType`.

Reply:
@smallx: Oh, I missed that. Thanks!